Python/codecs.c

```diff
@@ -1463,45 +1488,74 @@ static int _PyCodecRegistry_Init(void)
         }
     };
 
-    PyInterpreterState *interp = _PyInterpreterState_GET();
-    PyObject *mod;
-
-    if (interp->codec_search_path != NULL)
+    if (FT_ATOMIC_LOAD_INT_ACQUIRE(interp->codecs.initialized) == 1) {
         return 0;
-
-    interp->codec_search_path = PyList_New(0);
-    if (interp->codec_search_path == NULL) {
-        return -1;
     }
 
-    interp->codec_search_cache = PyDict_New();
-    if (interp->codec_search_cache == NULL) {
-        return -1;
+    PyObject *search_path = NULL, *search_cache = NULL, *error_registry = NULL;
```

Comment:
I think it might be marginally simpler, but why not use a `_PyOnceFlag` for initialization?

Reply:
Oh yeah, I remember now that we had talked about `_PyOnceFlag`. I'll give it a try.

Reply:
I was going to suggest `_PyOnceFlag`, but I think we'll want to synchronize `PyCodec_Unregister` with initialization and that's harder to do with a `_PyOnceFlag`.

If you get it to work well with `_PyOnceFlag`, then go for it.

Reply:
> We should only skip the import of encodings if we've already successfully imported that module.

How strictly should I follow this? The old version of this code would only try to import `encodings` once. If that failed, that specific call to `_PyCodecRegistry_Init()` would return -1, but subsequent calls into this file would think initialization was complete. [This line](https://github.com/python/cpython/blob/6258844c27e3b5a43816e7c559089a5fe0a47123/Python/codecs.c#L1505) after the import is misleading; nothing reads `codecs_initialized` in the old code.

I matched that behavior (for `encodings`, anyway. the new code will avoid setting any of the interpreter-global state if we fail to create part of it). If you're suggesting that I instead have subsequent calls to `_PyCodecRegistry_EnsureInit()` retry the import if it failed last time, I can do that, but it will require another piece of state tracking whether the import succeeded. It would be a visible behavioral change, although only in error cases.

Reply:
OK, yeah maybe don't change it. I'm looking at how this is actually called and it happens pretty early in interpreter initialization:

```
init_interp_main -> _PyUnicode_InitEncodings -> init_fs_encoding -> config_get_codec_name -> _PyCodec_Lookup -> _PyCodecRegistry_Init -> ... -> PyCodec_Register
```

I think we could simplify this all by explicitly initializing the codec registry instead of sort-of-lazily initializing it, but that would be a more involved change than you PR.